Update cosma recipe

mtaillefumier opened the pull request:
- Update cosma recipe
- Update costa recipe
- Add tiled-mm recipe (cosma hard dependency when CUDA/ROCM is used)
- Add semiprof : optional dependency

authors : @simonpintarelli , @mtaillefumier

Commit:
Update cosma recipe

Signed-off-by: Dr. Mathieu Taillefumier <mathieu.taillefumier@free.fr>

Commit:
Add maintainers list to tiled-mm

- fix typos
- updated maintainers list of cosma and costa

Commit:
Fix a typo in costa options

RMeli commented on var/spack/repos/builtin/packages/cosma/package.py:
```diff
@@ -12,14 +12,25 @@ class Cosma(CMakePackage):
     """
 
-    maintainers("haampie", "kabicm", "teonnik")
+    maintainers = ["haampie", "kabicm", "teonnik", "simonpintarelli", "mtaillefumier"]
```
```suggestion
    maintainers("haampie", "kabicm", "teonnik", "simonpintarelli", "mtaillefumier")
```

RMeli commented on var/spack/repos/builtin/packages/costa/package.py:
```diff
@@ -13,32 +13,35 @@ class Costa(CMakePackage):
-    maintainers("haampie", "kabicm", "RMeli")
+    maintainers("haampie", "kabicm", "RMeli", "mtaillefumier")
     homepage = "https://github.com/eth-cscs/COSTA"
+    url = "https://github.com/eth-cscs/COSTA/releases/download/v2.0/COSTA-v2.0.tar.gz"
```
This was removed in #35137 in favour of `url_from_version` (see https://github.com/spack/spack/pull/35137#discussion_r1085678357).

RMeli replied:
```suggestion
```

RMeli commented on var/spack/repos/builtin/packages/costa/package.py:
```diff
@@ -13,32 +13,35 @@ class Costa(CMakePackage):
 
     def url_for_version(self, version):
-        if version == Version("2.0"):
+        if version <= Version("2.0"):
```
```suggestion
        if version == Version("2.0"):
```
Even if there is no version `1.0` in this package, version `2.0` is the only one with the release artefact (i.e. this would break with version `1.0`).

mtaillefumier replied:
it should be fixed now

Commit:
Follow suggestions from reviewers

mtaillefumier committed:
Merge branch 'develop' into cosma_update

simonpintarelli commented:
`cosma_blas_cmake_arg` should be changed to the following:
```
    def cosma_blas_cmake_arg(self):
        # order matters, CUDA and ROCM need to be first,
        # since costa might introduce ^intel-mkl (scalapack provider)
        query_to_cmake_arg = [
            ("+cuda", "CUDA"),
            ("+rocm", "ROCM"),
            ("^intel-mkl", "MKL"),
            ("^intel-oneapi-mkl", "MKL"),
            ("^cray-libsci", "CRAY_LIBSCI"),
            ("^netlib-lapack", "CUSTOM"),
            ("^openblas", "OPENBLAS")
        ]
```

Otherwise `COSMA_BLAS` might get set to mkl for `cosma+cuda`, since `costa+scalapack` might introduce `^intel-mkl` (scalapack provider).

Commit:
Change order of BLAS implementations

mtaillefumier commented:
Would it be possible to merge this PR if there is nothing blocking it.

mtaillefumier committed:
Merge branch 'develop' into cosma_update

simonpintarelli reviewed (dismissed):
lgtm

alalazo commented on var/spack/repos/builtin/packages/cosma/package.py:
```diff
@@ -12,14 +12,30 @@ class Cosma(CMakePackage):
+    version(
+        "2.6.3",
+        sha256="8ca96ca41458f1e9d0da70d524c5a03c677dba7238d23a578f852163b6d45ac9",
+        url="https://github.com/eth-cscs/COSMA/releases/download/v2.6.3/cosma.tar.gz",
```
Can we update the `url` at the package level (which seems consistent for newer versions) and treat old versions as the special case?

mtaillefumier replied:
I will take care of it

mtaillefumier replied:
it is fixed in the last commit. Thanks for pointing that out.

alalazo reviewed (changes requested):
One minor request

Commit:
More version update

kabicm reviewed (dismissed):
Thanks Mathieu, looks good!

Commit:
Updated cosma archive checksum and costa version

- updated cosma version (in the cosma build system)
- updated costa version
- use the default generic url for downloading packages
- do not build tiled-mm when the cpu only version is needed

simonpintarelli reviewed (approved):
lgtm

RMeli commented on var/spack/repos/builtin/packages/cosma/package.py:
```diff
@@ -34,37 +40,50 @@ class Cosma(CMakePackage):
+        depends_on("tiled-mm@2.1+rocm", when="+rocm")
+        depends_on("tiled-mm@2.1+cuda", when="+cuda")
+
+    with when("@2.6.1:")
```
```suggestion
    with when("@2.6.1:"):
```

RMeli replied:
Fixes
```
==> Error: invalid syntax (package.py, line 67)
```

RMeli commented on var/spack/repos/builtin/packages/cosma/package.py:
```diff
@@ -34,37 +40,50 @@ class Cosma(CMakePackage):
-        return "https://github.com/eth-cscs/COSMA/releases/download/v{0}/COSMA-v{1}.tar.gz".format(
-            version, version
-        )
+    with when("@2.6.1:@2.6.2"):
```
```suggestion
    with when("@2.6.1:2.6.2"):
```

RMeli replied:
Fixes
```
==> Error: @2.6.1: cannot have multiple versions
```

RMeli commented on var/spack/repos/builtin/packages/tiled-mm/package.py:
```diff
@@ -0,0 +1,45 @@
+    depends_on("cxxopts", when="+examples")
+
+    conflicts("~cuda~rocm")
+
```
```suggestion
    conflicts("+cuda", when="+rocm")

```

mtaillefumier committed:
Update var/spack/repos/builtin/packages/tiled-mm/package.py

Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

mtaillefumier committed:
Update var/spack/repos/builtin/packages/cosma/package.py

Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

mtaillefumier committed:
Update var/spack/repos/builtin/packages/cosma/package.py

Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

mtaillefumier committed:
Merge branch 'develop' into cosma_update

Commit:
tiled-mm versions are incosistent

haampie committed:
Merge branch 'develop' into cosma_update
